src/args.rs

```diff
@@ -0,0 +1,441 @@
+	///
+	/// [THP]: https://www.kernel.org/doc/html/latest/admin-guide/mm/transhuge.html
+	#[clap(long)]
+	#[merge(strategy = merge::option::overwrite_none)]
+	#[cfg(target_os = "linux")]
+	pub thp: Option<bool>,
+
+	/// Kernel Samepage Merging
+	///
+	/// Advise the kernel to enable Kernel Samepage Merging [KSM] on the virtual RAM.
+	///
+	/// [KSM]: https://www.kernel.org/doc/html/latest/admin-guide/mm/ksm.html
+	#[clap(long)]
+	#[merge(strategy = merge::option::overwrite_none)]
+	#[cfg(target_os = "linux")]
+	pub ksm: Option<bool>,
+}
+
+/// Arguments for the CPU resources allocated to the guest.
+#[derive(Default, Debug, Clone, Merge, Parser, Deserialize)]
+pub struct CpuArgs {
+	/// Number of guest CPUs
+	#[clap(short, long, env = "HERMIT_CPU_COUNT")]
+	#[merge(strategy = merge::option::overwrite_none)]
+	pub cpu_count: Option<CpuCount>,
+
+	/// Create a PIT
+	#[clap(long)]
+	#[merge(strategy = merge::option::overwrite_none)]
+	#[cfg(target_os = "linux")]
+	pub pit: Option<bool>,
+
+	/// Bind guest vCPUs to host cpus
+	///
+	/// A list of host CPU numbers onto which the guest vCPUs should be bound to obtain performance benefits.
+	/// List items may be single numbers or inclusive ranges.
+	/// List items may be separated with commas or spaces.
+	///
+	/// # Examples
+	///
+	/// * `--affinity "0 1 2"`
+	///
+	/// * `--affinity 0-1,2`
+	#[clap(short, long, name = "CPUs")]
+	#[merge(strategy = merge::option::overwrite_none)]
+	pub affinity: Option<Affinity>,
+}
+
+#[derive(Debug, Clone)]
+pub struct Affinity(Vec<CoreId>);
+
+impl Affinity {
+	pub fn parse_ranges_iter<'a>(
+		ranges: impl IntoIterator<Item = &'a str> + 'a,
+	) -> impl Iterator<Item = Result<usize, ParseIntError>> + 'a {
+		struct ParsedRange(RangeInclusive<usize>);
+
+		impl FromStr for ParsedRange {
+			type Err = ParseIntError;
+
+			fn from_str(s: &str) -> Result<Self, Self::Err> {
+				let range = match s.split_once('-') {
+					Some((start, end)) => start.parse()?..=end.parse()?,
+					None => {
+						let idx = s.parse()?;
+						idx..=idx
+					}
+				};
+				Ok(Self(range))
+			}
+		}
+
+		ranges
+			.into_iter()
+			.map(ParsedRange::from_str)
+			.flat_map(|range| match range {
+				Ok(range) => Either::Left(range.0.map(Ok)),
+				Err(err) => Either::Right(iter::once(Err(err))),
+			})
+	}
+
+	fn parse_ranges(ranges: &str) -> Result<Vec<usize>, ParseIntError> {
+		Self::parse_ranges_iter(ranges.split([' ', ','].as_slice())).collect()
+	}
+}
+
+#[derive(Error, Debug)]
+pub enum ParseAffinityError {
+	#[error(transparent)]
+	Parse(#[from] ParseIntError),
+
+	#[error("Available cores: {available_cores:?}, requested affinities: {requested_affinities:?}")]
+	InvalidValue {
+		available_cores: Vec<usize>,
+		requested_affinities: Vec<usize>,
+	},
+}
+
+impl FromStr for Affinity {
+	type Err = ParseAffinityError;
+
+	fn from_str(s: &str) -> Result<Self, Self::Err> {
+		let available_cores = core_affinity::get_core_ids()
```

Comment:
You need to provide a way to mock/overwrite this for tests, because otherwise they depend on the core count of the test host, which is bad.

Reply:
I will not address this in this PR - I would like to maintain the functionality of being able to show why the user's input is problematic.